test/SILGen/back_deploy_attribute.swift

```diff
@@ -1,36 +1,101 @@
+  // CHECK: [[RESULT:%.*]] = apply [[FALLBACKFN]]([[BB0_ARG]]) : $@convention(method) (TopLevelStruct) -> ()
+  // CHECK: br [[UNAVAIL_RESULT_BB:bb[0-9]+]]([[RESULT]] : $())
+  //
+  // CHECK: [[UNAVAIL_RESULT_BB]]([[UNAVAIL_RESULT_BB_ARG:%.*]] : $()):
+  // CHECK: br [[RETURN_BB:bb[0-9]+]]([[UNAVAIL_RESULT_BB_ARG]] : $())
+  //
+  // CHECK: [[AVAIL_BB]]:
+  // CHECK: [[ORIGFN:%.*]] = function_ref @$s21back_deploy_attribute14TopLevelStructV0a8DeployedF4FuncyyF : $@convention(method) (TopLevelStruct) -> ()
+  // CHECK: [[RESULT:%.*]] = apply [[ORIGFN]]([[BB0_ARG]]) : $@convention(method) (TopLevelStruct) -> ()
+  // CHECK: br [[RESULT_BB:bb[0-9]+]]([[RESULT]] : $())
+  //
+  // CHECK: [[RESULT_BB]]([[RESULT_BB_ARG:%.*]] : $()):
+  // CHECK: br [[RETURN_BB]]([[RESULT_BB_ARG]] : $())
+  //
+  // CHECK: [[RETURN_BB]]([[RETURN_BB_ARG:%.*]] : $())
+  // CHECK: return [[RETURN_BB_ARG]] : $()
+
+  // ---- Original definition of TopLevelStruct.backDeployedStructFunc()
   // CHECK-LABEL: sil [serialized] [available 10.51] [ossa] @$s21back_deploy_attribute14TopLevelStructV0a8DeployedF4FuncyyF : $@convention(method) (TopLevelStruct) -> ()
   @available(macOS 10.51, *)
   @_backDeploy(macOS 10.50)
   public func backDeployedStructFunc() {}
 }
 
-
-// FIXME(backDeploy): Verify SIL in a caller that requires back deployment
-
-
-// CHECK-LABLEL: sil hidden [available 10.51] [ossa] @$s21back_deploy_attribute21alwaysAvailableCalleryyAA14TopLevelStructVF : $@convention(thin) (TopLevelStruct) -> ()
-// CHECK: bb0(%0 : $TopLevelStruct):
+// CHECK-LABEL: sil hidden [available 10.51] [ossa] @$s21back_deploy_attribute6calleryyAA14TopLevelStructVF : $@convention(thin) (TopLevelStruct) -> ()
 @available(macOS 10.51, *)
-func alwaysAvailableCaller(_ s: TopLevelStruct) {
-  /// This function's availability meets the minimum availability of the APIs, so
-  /// no back deployment logic is required.
-
-  // CHECK: %2 = function_ref @$s21back_deploy_attribute0A12DeployedFuncyyF : $@convention(thin) () -> ()
-  // CHECK: %3 = apply %2() : $@convention(thin) () -> ()
+func caller(_ s: TopLevelStruct) {
+  // CHECK: bb0([[STRUCT_ARG:%.*]] : $TopLevelStruct):
+  // CHECK: [[FNREF:%.*]] = function_ref @$s21back_deploy_attribute0A12DeployedFuncyyFTwb : $@convention(thin) () -> ()
+  // CHECK: [[FNRES:%.*]] = apply [[FNREF]]() : $@convention(thin) () -> ()
   backDeployedFunc()
-  // CHECK: %4 = function_ref @$s21back_deploy_attribute14TopLevelStructV0a8DeployedF4FuncyyF : $@convention(method) (TopLevelStruct) -> ()
-  // CHECK: %5 = apply %4(%0) : $@convention(method) (TopLevelStruct) -> ()
+  // CHECK: [[FNREF:%.*]] = function_ref @$s21back_deploy_attribute14TopLevelStructV0a8DeployedF4FuncyyFTwb : $@convention(method) (TopLevelStruct) -> ()
+  // CHECK: [[FNRES:%.*]] = apply [[FNREF]]([[STRUCT_ARG]]) : $@convention(method) (TopLevelStruct) -> ()
   s.backDeployedStructFunc()
 }
```

Comment:
Suggestions for more tests:

- Complex parameter lists (indirectly-passed values like existentials, inout parameters, tuples)
- Complex returns (indirectly-returned values like existentials, tuples)
- Generic functions
- Async and throws functions
- Anything with a non-trivial body

Also please add tests in test/attr/ or test/Sema/ to ensure that inlinable restrictions are enforced in the body of a @_backDeploy just like they are for @_aeic.

Reply:
Good point, I should cover many more types of functions. One thing I wonder about is the scalability of testing all those different function types as SILGen tests. Do you think it's important that we verify the SIL for each of those configurations, or would it be satisfactory to have a validation test that calls many different types of back deployed functions and verifies their behaviors? I was planning on writing the latter kind of test in upcoming PRs after a bit more of the type checking is in place.

I do have an existing test in `/test/attr` verifying the type checking diagnostics for this attribute. However, much of the type checking is still coming in later PRs. I wanted to get this working end to end to make sure I understood how it would work before adding all the polish.

Reply:
I don't think it makes sense to cover every possible case, I was just thinking of the high-value ones that might identify bugs in how you're forwarding arguments and results.

Reply:
Sounds good, I'll cover what you listed.

Reply:
Testing generics revealed that the thunk is not quite forwarding arguments perfectly for all functions correctly yet so I'm continuing to investigate that.

Reply:
I now have tests covering back deployment of the following function archetypes:
- A trivial top level function with no arguments or results.
- A top level function that takes an argument and returns a value.
- A trivial top level function that throws.
- A trivial top level function on a struct.
- A generic function that takes and returns a `T`.

I also experimented with async functions, functions taking `inout` params, and functions taking/returning an existential but did not end up writing any SILGen test for those specifically because they didn't appear to me to exercise distinct logic in my SILGen code. I think there may be diminishing returns on more SILGen tests. However, I do intend to write executable validation tests that exercise a wider variety of function archetypes in an upcoming PR.